core/dbt/parser/unit_tests.py

```diff
@@ -131,6 +131,7 @@ def parse_unit_test_case(self, test_case: UnitTestDefinition):
                 "fqn": original_input_node.fqn,
                 "checksum": FileHash.empty(),
                 "raw_code": self._build_fixture_raw_code(given.rows, None),
+                "defer_relation": original_input_node.defer_relation,
```

Comment:
This works in combination with a change to `get_fixture_sql` that prefers to get column types from `defer_relation` if available, otherwise falling back to `this`

https://github.com/dbt-labs/dbt-adapters/pull/60